### PR TITLE
Changes to allow pmap-of-sharded_jit to run on multi-host platforms.

### PR DESCRIPTION
**Pull request by skye**
* Checks that we're running on all local devices after tracing, so we
  have the full shard count instead of just the axis size.

* Makes sure the device assignment is always the right shape (we
  currently generate 1D default device assignments for multi-host)

### CODE REVIEW

**Comment by skye** [comment]
@trevorcai any chance you wanna review this?

File: jax/interpreters/pxla.py
Comment (trevorcai) [inline jax/interpreters/pxla.py:641]: Probably don't want to lose the logging from jax_log_compiles, consider relocating to another spot.

File: jax/interpreters/pxla.py
Comment (trevorcai) [inline jax/interpreters/pxla.py:759]: At least in the nested pmap case, device_assignment order seems to matter a lot for all-reduce times.
This could be true for sharded_jit as well - worth keeping an eye out for this.

File: jax/interpreters/pxla.py
Comment (trevorcai) [inline jax/interpreters/pxla.py:622]: I think this is problematic when evaluating this logic on the pmap of nested cmap/pmap.

Consider evaluating the following on 2 hosts of 8 devices each.
```jax.pmap(jax.pmap(lambda x: x * 2, global_axis_size=4), global_axis_size=4)```
The inner, local, pmap axis will computed `expected = 4 * 2 = 8`, which does not match the specified `global_axis_size`.

The original (convoluted) control flow only infers`global_axis_size` from data & context when it's `None` for this reason. (Alternatively: examine the current AxisEnv and don't do this when inside a pmap.)

Is this logic problematic in the context of sharding?

File: jax/interpreters/pxla.py
Comment (trevorcai) [inline jax/interpreters/pxla.py:687]: Similarly, this check is also problematic for nested multihost pmaps.
On the same cmap/pmap example as above, we get for the inner pmap:
```
num_local_replicas = 4 * 1 = 4
num_local_shards = 4 * 1 = 4
num_local_shards != xb.local_device_count()
```

I'm not sure how to fix this chunk, but I think the reasonable thing to do is to say must_run_on_all_devices if global_axis_size is specified (and optionally is less than local_device_count).

**Review by trevorcai — COMMENTED** [review]
I upstreamed some basic nested pmap stuff into your experimental script, it might be worth running against that if you have the chance.

File: jax/interpreters/pxla.py
Comment (skye) [reply jax/interpreters/pxla.py]: Blah you're right. I forgot this code is run for inner pmaps, so I was only considering the outermost case. I put back most of the convoluted original logic, and also check the DynamicAxisEnv to only use `axis_size * host_count` for the outermost pmap if there's no user-provided `global_axis_env`.

I also made it an error if no `global_axis_size` is specified for an inner pmap. This should be an error right?

This works for sharding, because inner sharded_jits go through a simpler non-multihost-aware path.

(BTW, the only reason this runs for inner pmaps is to do partial evaluation, i.e. to evaluate any non-data-dependent values in the inner pmap, which is gonna go away soon. Then this code will only run for the outermost pmap, which will be much simpler to reason about.)

File: jax/interpreters/pxla.py
Comment (skye) [reply jax/interpreters/pxla.py]: I expanded the conditional to only fire for outer pmaps. Everything should add up in the outer pmap. There might be errors we could check for in the inner pmap too, but this code won't be run for inner pmaps soon (see above comment).

File: jax/interpreters/pxla.py
Comment (skye) [reply jax/interpreters/pxla.py:765]: Agreed. I'd like to create a fancier "default device assignment" API that we can implement in XLA, but it's kind of hard to describe the constraints (something like, these replicas need to be on these hosts).

**Review by skye — COMMENTED** [review]
Thanks for the review! I ran my script and it now passes (it didn't before, derp).

**Comment by skye** [comment]
BTW, I'm gonna wait for your approval before submitting. In particular, I added an error if no `global_axis_size` is specified for inner pmaps but I'm not 100% sure about that :)

File: jax/interpreters/pxla.py
Comment (trevorcai) [reply jax/interpreters/pxla.py:765]: I'll send you the code I'm currently using for generating device_assignment ordering (and the too-loose constraints it's using), in case it's helpful.

File: jax/interpreters/pxla.py
Comment (trevorcai) [inline jax/interpreters/pxla.py:615]: I'm not too familiar with what's going on in DynamicAxisEnv so please override me, but isn't this reversed?
We're in an inner pmap if there are already items in `dynamic_axis_env`; i.e. len > 0?

File: jax/interpreters/pxla.py
Comment (trevorcai) [reply jax/interpreters/pxla.py]: This seems like a reasonable fix.
Is this yet another problem that omnistaging is going to solve? Y'all are making it sound like the greatest thing since sliced bread :)

File: jax/interpreters/pxla.py
Comment (trevorcai) [reply jax/interpreters/pxla.py:641]: Think you left it out of the commit, not seeing the new logging.

**Review by trevorcai — COMMENTED** [review]
Looks good overall, just some minor questions.

File: jax/interpreters/pxla.py
Comment (skye) [reply jax/interpreters/pxla.py]: Gah you're right! I guess it still worked because the test script has `global_axis_size` set for both pmaps, and I'm guessing the inner pmap never makes it to the `num_local_shards` check because partial eval probably yields a trivial computation that isn't actually compiled. I fixed it, and FWIW the test script still passes...

File: jax/interpreters/pxla.py
Comment (skye) [reply jax/interpreters/pxla.py:641]: Oof. I'm copy/pasting internal/external code back and forth, and keep screwing up this logging! (I actually tried to move it from the start but kept failing somehow.) I think it's actually there now...

File: jax/interpreters/pxla.py
Comment (skye) [reply jax/interpreters/pxla.py:771]: Thanks, I think it will be!

**Review by trevorcai — APPROVED** [review]

**Comment by skye** [comment]
Thanks again for the review!

File: jax/interpreters/pxla.py
Comment (jekbradbury) [reply jax/interpreters/pxla.py]: Yes, yes it is!

**Pull request merged**
